Feature/create summaries

austin007008 opened the pull request:
adds new functionality to create summaries for returned documents

austin007008 committed:
add toption to return summaries for returned documents in querys

working summaries created

fix test

reccomendations

trailing regex for name matching

remove unused import

parameters can be entered in any order

apmoriarty commented on warehouse/query-core/src/main/java/datawave/query/QueryParameters.java:
```diff
@@ -211,6 +211,11 @@ public class QueryParameters {
     public static final String EXCERPT_FIELDS = "excerpt.fields";
 
+    /**
+     * Used to specify summaries that should be returned.
```
Consider renaming to SUMMARY_SIZE. Doc should be updated to include the size unit and any default value (I think the default value is 150 below). Size in bytes? The number of words or tokens?

austin007008 replied:
fixed

apmoriarty commented on warehouse/query-core/src/main/java/datawave/query/attributes/SummarySize.java:
```diff
@@ -0,0 +1,215 @@
+    private boolean only;
+
+    public SummarySize() {
+        summarySize = 0;
```
This is a bit confusing. The default summary size is 150, but the default constructor sets the size to zero.

austin007008 replied:
currently the default size is used if the input to the function is misformed

apmoriarty commented on warehouse/query-core/src/main/java/datawave/query/planner/DefaultQueryPlanner.java:
```diff
@@ -655,6 +658,11 @@ private void configureExcerpts(ShardQueryConfiguration config, IteratorSetting c
         }
     }
 
+    private void configureSummaries(ShardQueryConfiguration config, IteratorSetting cfg) {
```
This should only be added if summaries are requested. Wrap in something like 
```if (config.getSummarySize() > 0) {```

austin007008 replied:
fixed

apmoriarty commented on warehouse/query-core/src/main/java/datawave/query/planner/DefaultQueryPlanner.java:
```diff
@@ -1868,6 +1876,11 @@ protected ASTJexlScript upperCaseIdentifiers(MetadataHelper metadataHelper, Shar
             Sets.newHashSet(excerptFields.getFields()).stream().forEach(s -> excerptFields.replace(s, s.toUpperCase()));
         }
 
+        SummarySize summarySize = config.getSummarySize();
+        if (summarySize != null && !summarySize.isEmpty()) {
+            summarySize.getContentNames().stream().forEach(s -> summarySize.replace(s, s.toUpperCase()));
+        }
```
Uppercasing the content names should happen internally to SummarySize anytime new names are added. 

Edit: I see you're doing the same thing as ExcerptFields, but ExceprtFields should also be enforcing upper case internally instead of relying on an external class.

austin007008 replied:
fixed

drewfarris committed:
Updated AccumuloSyntaxParser to accept functions with no arguments

Re-generated AccumuloSyntaxParser.java from AccumuloSyntaxParser.jj using
a verson of javacc 4.1 build from https://github.com/javacc/javacc/tree/release_41

austin007008 committed:
can take an empty function for default values

austin007008 committed:
make query parameter SUMMARY instead of SUMMARY_SIZE

clean

apmoriarty commented on warehouse/query-core/src/main/java/datawave/query/attributes/SummarySize.java:
```diff
@@ -0,0 +1,201 @@
+ * Represents options for a summary that have been specified within an #SUMMARY_SIZE function. An instance of {@link SummarySize} can easily be captured as a
+ * parameter string using {@link SummarySize#toString()}, and transformed back into a {@link SummarySize} instance via {@link SummarySize#from(String)}.
+ */
+public class SummarySize implements Serializable {
```
Rename to "SummaryOptions"

austin007008 replied:
done

apmoriarty commented on warehouse/query-core/src/main/java/datawave/query/attributes/SummarySize.java:
```diff
@@ -0,0 +1,201 @@
+                String[] parts = parameterPart.split(Constants.COLON);
+                if (parts[0].equalsIgnoreCase(SIZE_PARAMETER)) {
+                    summarySize.summarySize = Integer.parseInt(parts[1]);
+                } else if (parts[0].equalsIgnoreCase("ONLY")) {
```
extract "ONLY" to private constant?

austin007008 replied:
done

apmoriarty commented on warehouse/query-core/src/main/java/datawave/query/language/functions/jexl/SummarySize.java:
```diff
@@ -0,0 +1,57 @@
+ * {@code size/[only]/[contentName1, contentName2, ....]}. See {@link datawave.query.attributes.SummarySize} for additional documentation on supported
+ * formatting.
+ */
+public class SummarySize extends JexlQueryFunction {
```
No lucene version?

austin007008 replied:
all of the lucene functions have been deprecated since 2021

apmoriarty commented on warehouse/query-core/src/main/java/datawave/query/tables/ShardQueryLogic.java:
```diff
@@ -1537,6 +1546,26 @@ public void setExcerptIterator(String iteratorClass) {
+        } catch (Exception e) {
+            throw new DatawaveFatalQueryException("Illegal d column summary iterator class", e);
+        }
+    }
```
This is actually the iterator class name. Follow the pattern used by the create uids iterator, i.e., use `getCreateUidsIteratorClass` and `setCreateUidsIteratorClass`

austin007008 replied:
I think it is like this so that the specified class can be easily changed (could be wrong about that though). See set/get ExcerptIterator

apmoriarty commented on warehouse/query-core/src/main/java/datawave/query/transformer/SummaryTransform.java:
```diff
@@ -0,0 +1,274 @@
+
+    private static final Logger log = LoggerFactory.getLogger(SummaryTransform.class);
+
+    public static final String SUMMARY_ERROR_MESSAGE = "SOMETHING WENT WRONG GENERATING YOUR SUMMARY!";
```
This can be simplified to "UNABLE TO GENERATE SUMMARY"

austin007008 replied:
done

apmoriarty commented on warehouse/query-core/src/main/java/datawave/query/transformer/SummaryTransform.java:
```diff
@@ -0,0 +1,274 @@
+import datawave.query.iterator.logic.DColumnSummaryIterator;
+import datawave.query.iterator.logic.TermFrequencyExcerptIterator;
+
+public class SummaryTransform extends DocumentTransform.DefaultDocumentTransform {
```
Class level documentation please

austin007008 replied:
done

apmoriarty commented on warehouse/query-core/src/main/java/datawave/query/transformer/SummaryTransform.java:
```diff
@@ -0,0 +1,274 @@
+     *            the pre-identified document keys
+     * @return the excerpts
+     */
+    private Set<Summary> getExcerpts(final ArrayList<DocumentKey> documentKeys) {
```
Rename to "getSummaries"

austin007008 replied:
done

apmoriarty commented on warehouse/query-core/src/test/java/datawave/query/util/SummaryTest.java:
```diff
@@ -0,0 +1,321 @@
+
+        String queryString = "QUOTE:(farther) #SUMMARY(SIZE:-50)";
+
+        // not sure why the timestamp and delete flag are present
```
did we figure out why the timestamp and delete flag are present?

drewfarris replied:
The timestamp and delete flag will always be present as components of the key if we're using `toString()`, won't they?

austin007008 replied:
idk tbh. the layout of this test was taken from "ExcerptTest" and it had the same thing.

apmoriarty commented on warehouse/query-core/src/test/java/datawave/query/util/SummaryTest.java:
```diff
@@ -0,0 +1,321 @@
+        extraParameters.put("return.fields", "CONTENT_SUMMARY");
+        extraParameters.put("query.syntax", "LUCENE");
+
+        String queryString = "QUOTE:(farther) #SUMMARY(SIZE:50/ONLY/VIEWS:CANTFINDME,ORME)";
```
I would like to see some more tests surrounding the configuration so there are concrete examples of how the iterator behaves with different inputs. 

It seems odd that if I specify a view name that is additive to some default view names. 

You should not have to specify 'CONTENT_SUMMARY' in the return fields in order to get a summary back. So really you need to handle two distinct cases
1) user does not specify 'return.fields', nothing to do
2) user specifies 'return.fields', need to add 'CONTENT_SUMMARY'

austin007008 replied:
1.  will try and add more tests
2. incase the default ones do not cover everything and it also allows you to change the priority
3. you dont. it seems that is just for the test so we only get back the fields we care about (see ExcerptTest).

austin007008 replied:
done

apmoriarty commented on warehouse/query-core/src/main/java/datawave/query/transformer/SummaryTransform.java:
```diff
@@ -0,0 +1,273 @@
+     * @param document
+     *            the document
+     */
+    private static void addExcerptsToDocument(Set<Summary> summaries, Document document) {
```
rename to 'addSummariesToDocument'

austin007008 replied:
done

drewfarris commented on warehouse/query-core/src/main/java/datawave/query/attributes/SummaryOptions.java:
```diff
@@ -0,0 +1,213 @@
+
+    private int summarySize;
+    private ArrayList<String> viewNamesList;
+    private boolean only;
```
consider renaming this variable to `onlyListedViews`. 

Could you explain a bit more in here what this variable does?

austin007008 replied:
done

drewfarris commented on warehouse/query-core/src/main/java/datawave/query/iterator/logic/DColumnSummaryIterator.java:
```diff
@@ -0,0 +1,384 @@
+/**
+ * This iterator is intended to scan the d column for a specified document. The result will be a summary for each document scanned.
+ */
+public class DColumnSummaryIterator implements SortedKeyValueIterator<Key,Value> {
```
Consider renaming this something like 'ContentSummaryIterator' - elsewhere we don't really call anything 'DColumn(whatever)', but call things related to the d columne 'Content(whatever)'.

austin007008 replied:
done

drewfarris commented on warehouse/query-core/src/main/java/datawave/query/iterator/logic/DColumnSummaryIterator.java:
```diff
@@ -0,0 +1,384 @@
+     *            the size in bytes of the summary to create
+     * @return the created summary
+     */
+    private static String createSummary(List<String> viewSummaryOrder, Map<String,byte[]> foundContent, int summarySize) {
```
Consider making this logic pluggable? Perhaps we'll have a different summarization algorithm to add later? The rest of the class is concerned with retrieving the content to summarize, it might be helpful to have the summarization performed in a separate class.

Less ideal, but minimally consider making this protected so that a subclass could implement something different?

austin007008 replied:
done

drewfarris commented on warehouse/query-core/src/test/java/datawave/query/iterator/logic/DColumnSummaryIteratorTest.java:
```diff
@@ -0,0 +1,255 @@
+import datawave.query.iterator.SortedListKeyValueIterator;
+
+@RunWith(EasyMockRunner.class)
+public class DColumnSummaryIteratorTest extends EasyMockSupport {
```
As with the iterator, consider renaming this to `ContentSummaryIteratorTest`

austin007008 replied:
done

drewfarris commented:
Looks like the builds might be failing due to a formatting issue - consider double checking this.

austin007008 committed:
updates pt1

austin007008 committed:
recommendation pt2

FineAndDandy commented on warehouse/query-core/src/main/java/datawave/query/attributes/SummaryOptions.java:
```diff
@@ -0,0 +1,229 @@
+            }
+        } catch (Exception e) {
+            log.warn("Unable to parse summary size string, returning empty SummaryOptions: {}", string, e);
+            return new SummaryOptions();
```
maybe this should be a fatal rather that default situation?

FineAndDandy commented on warehouse/query-core/src/main/java/datawave/query/iterator/logic/ContentSummaryIterator.java:
```diff
@@ -0,0 +1,329 @@
+    public void init(SortedKeyValueIterator<Key,Value> source, Map<String,String> options, IteratorEnvironment env) throws IOException {
+        this.source = source;
+
+        viewSummaryOrder.addAll(List.of("CONTENT", "CONTENT1"));
```
Should this include CONTENT1?

austin007008 replied:
done

FineAndDandy commented on warehouse/query-core/src/main/java/datawave/query/iterator/logic/ContentSummaryIterator.java:
```diff
@@ -0,0 +1,329 @@
+        viewSummaryOrder.addAll(List.of("CONTENT", "CONTENT1"));
+
+        if (options.containsKey(SUMMARY_SIZE)) {
+            this.summarySize = Math.max(1, Math.min(Integer.parseInt(options.get(SUMMARY_SIZE)), 1500));
+        } else {
+            this.summarySize = 150;
```
define constants

austin007008 replied:
done

FineAndDandy commented on warehouse/query-core/src/main/java/datawave/query/iterator/logic/ContentSummaryIterator.java:
```diff
@@ -0,0 +1,329 @@
+     *             for issues with read/write
+     */
+    private void seekToNextUid(Text row, String dtAndUid) throws IOException {
+        Key startKey = new Key(row, Constants.D_COLUMN_FAMILY, new Text(dtAndUid + '.'));
```
is a '.' the right character here? the dtAndUid will be followed by a null-byte right?

austin007008 replied:
done

FineAndDandy commented on warehouse/query-core/src/main/java/datawave/query/iterator/logic/ContentSummaryIterator.java:
```diff
@@ -0,0 +1,329 @@
+                if (name.endsWith("*")) {
+                    name = name.substring(0, name.length() - 1);
+                    if (currentViewName.startsWith(name)) {
+                        foundContent.put(currentViewName, source.getTopValue().get());
+                    }
+                } else {
+                    if (currentViewName.equalsIgnoreCase(name)) {
+                        foundContent.put(currentViewName, source.getTopValue().get());
```
should there be a limit on the amount of content that will be added to prevent potential memory problems? This might also require some more configuration flags

drewfarris replied:
This is an interesting point and one I had on my mind too. 

if we just do a `getTopValue().get()` - we're just passing around a reference to a byte[] array that's already loaded into memory as a result of Accumulo reading it from an RFile - in some ways it's too late at that point. If we do some sort of truncation here, we'll end up creating a truncated copy of the original byte array, which consumes additional memory but should open up the original Value for GC.

Happy to punt on this for now and explore this as a optimization to add later. 

I really wish I could tell Accumulo to  not read the entire value in the first place.

austin007008 replied:
current implementation is if the content is over X size, it will immediately decode/decompress and truncate and save that smaller value vs saving the whole content and dealing with it later

FineAndDandy commented on warehouse/query-core/src/main/java/datawave/query/tables/ShardQueryLogic.java:
```diff
@@ -1537,6 +1546,26 @@ public void setExcerptIterator(String iteratorClass) {
         }
     }
 
+    public SummaryOptions getSummaryoptions() {
```
casing

austin007008 replied:
done

FineAndDandy commented on warehouse/query-core/src/main/java/datawave/query/tables/ShardQueryLogic.java:
```diff
@@ -988,6 +989,14 @@ protected void loadQueryParameters(ShardQueryConfiguration config, Query setting
+        String summaryParam = settings.findParameter(QueryParameters.SUMMARY_OPTIONS).getParameterValue().trim();
+        if (StringUtils.isNotBlank(summaryParam)) {
+            SummaryOptions summaryOptions = SummaryOptions.from(summaryParam);
+            this.setSummaryOptions(summaryOptions);
+            config.setSummaryOptions(summaryOptions);
```
should the config be updated in the call to setSummaryOptions()?

FineAndDandy commented on warehouse/query-core/src/main/java/datawave/query/tables/ShardQueryLogic.java:
```diff
@@ -1537,6 +1546,26 @@ public void setExcerptIterator(String iteratorClass) {
+        getConfig().setSummaryOptions(summaryOptions);
+    }
+
+    public String getSummaryIterator() {
```
rename to getSummaryIteratorClassName()

austin007008 replied:
done

FineAndDandy commented on warehouse/query-core/src/main/java/datawave/query/tables/ShardQueryLogic.java:
```diff
@@ -1537,6 +1546,26 @@ public void setExcerptIterator(String iteratorClass) {
+        return getConfig().getSummaryIterator().getName();
+    }
+
+    public void setSummaryIterator(String iteratorClass) {
```
rename to setSummaryIteratorClassName

austin007008 replied:
done

FineAndDandy commented on warehouse/query-core/src/test/java/datawave/query/config/ShardQueryConfigurationTest.java:
```diff
@@ -469,6 +471,10 @@ public void setUp() throws Exception {
         defaultValues.put("excerptIterator", TermFrequencyExcerptIterator.class);
         updatedValues.put("excerptIterator", TermFrequencyIndexIterator.class);
+        defaultValues.put("summaryOptions", new SummaryOptions());
+        updatedValues.put("summaryOptions", SummaryOptions.from("50"));
```
reference default constant

austin007008 replied:
done

FineAndDandy commented on warehouse/query-core/src/main/java/datawave/query/language/parser/lucene/AccumuloSyntaxParser.java:
was there an uncommitted file that triggered the auto gen updates?

austin007008 committed:
fix downstream

austin007008 committed:
fixups and not pulling all the content all the time

FineAndDandy commented on warehouse/query-core/src/main/java/datawave/query/iterator/logic/ContentSummaryIterator.java:
```diff
@@ -272,6 +278,17 @@ public void next() throws IOException {
+        if (content.length < MAX_CONTENT_SIZE) {
+            foundContent.put(currentViewName + Constants.COLON + Boolean.TRUE, content);
+        } else {
+            content = ContentKeyValueFactory.decodeAndDecompressContent(content);
+            foundContent.put(currentViewName + Constants.COLON + Boolean.FALSE, content);
```
this will allow putting more than the MAX_CONTENT_SIZE into memory.

austin007008 committed:
pre truncate big content

drewfarris commented:
Looks like we have a unit test failure that needs to be addressed before merging here.

drewfarris committed:
Unit test fixup

drewfarris committed:
Fixed formatting violation
